src/software/thunderscope/robot_communication.py

```diff
@@ -100,27 +122,124 @@ def __init__(
+            self.current_network_config.referee_interface = (
+                referee_interface if not error else DISCONNECTED
+            )
+
+        if not self.is_setup_for_fullsystem:
+            self.robots_connected_to_fullsystem = {
+                robot_id for robot_id in range(MAX_ROBOT_IDS_PER_SIDE)
+            }
+
+            self.is_setup_for_fullsystem = True
+
+    def __setup_for_robot_communication(
+        self, robot_status_interface: str = "lo"
+    ) -> None:
         """
-        self.receive_ssl_wrapper = tbots_cpp.SSLWrapperPacketProtoListener(
-            SSL_VISION_ADDRESS,
-            SSL_VISION_PORT,
-            lambda data: self.__forward_to_proto_unix_io(SSL_WrapperPacket, data),
+        Set up senders and listeners for communicating with the robots
+
+        :param robot_status_interface: the interface to listen/send for robot status data. Ignored for sending
+        primitives if using radio
+        """
+        if (
+            robot_status_interface == self.current_network_config.robot_status_interface
+            or robot_status_interface == DISCONNECTED
+        ):
+            return
+
+        # Create the multicast listeners
+        self.receive_robot_status, error = tbots_cpp.createRobotStatusProtoListener(
+            self.multicast_channel,
+            ROBOT_STATUS_PORT,
+            robot_status_interface,
+            self.__receive_robot_status,
             True,
         )
 
-        self.receive_ssl_referee_proto = tbots_cpp.SSLRefereeProtoListener(
-            SSL_REFEREE_ADDRESS,
-            SSL_REFEREE_PORT,
-            lambda data: self.current_proto_unix_io.send_proto(Referee, data),
+        self.receive_robot_log, error = tbots_cpp.createRobotLogProtoListener(
+            self.multicast_channel,
+            ROBOT_LOGS_PORT,
+            robot_status_interface,
+            lambda data: self.__forward_to_proto_unix_io(RobotLog, data),
             True,
         )
 
-        self.robots_connected_to_fullsystem = {
-            robot_id for robot_id in range(MAX_ROBOT_IDS_PER_SIDE)
-        }
+        self.receive_robot_crash, error = tbots_cpp.createRobotCrashProtoListener(
+            self.multicast_channel,
+            ROBOT_CRASH_PORT,
+            robot_status_interface,
+            lambda data: self.current_proto_unix_io.send_proto(RobotCrash, data),
+            True,
+        )
+
+        # Create multicast senders
+        if self.enable_radio:
+            self.send_primitive_set = tbots_cpp.PrimitiveSetProtoRadioSender()
+        else:
+            self.send_primitive_set, error = tbots_cpp.createPrimitiveSetProtoUdpSender(
+                self.multicast_channel, PRIMITIVE_PORT, robot_status_interface, True
+            )
+
+        if error:
+            print(f"Error setting up robot status interface: {error}")
+
+        self.current_network_config.robot_status_interface = (
+            robot_status_interface if not error else DISCONNECTED
+        )
```

Comment:
the same error message variable is used for all listeners/senders, with the variable only being checked at the end. I think this would overwrite any previous error messages. 

Testing this on my PC, I can set `vision_interface` to my wifi interface (my wifi is turned off and Im using ethernet) and I see `DISCONNECTED` in the terminal, but when I set `robot_status_interface` to my wifi interface, it shows it as green and connected. I checked and it seems like none of the error messages are filled, so maybe it was okay, but I did get this error message:
`An unknown network error occurred when attempting to shutdown UDP socket for TbotsProto::RobotStatus. The boost system error is: system:107: Transport endpoint is not connected`

Though, I'm not sure what we should when we get an error for one of the UDP endpoints, and not the other ones. I think the priority should be to have PrimitiveSets being sent out, and if the others fail it's not as big of an issue...

Reply:
unfortunately with IPv6 addresses in boost sockets, they fail silently and we won't know if it's connected through the constructor.

the error you receive when you try to close the IPv6 socket only shows up when you try to close it. I tried for quite a while to try to use the boost API to recognize IPv6 fault connections but I couldn't find a way